### PR TITLE
feat(dataset) Add the query plan and QueryPlanBuilder

### PR DESCRIPTION
**Pull request by fpacifici**
This is the second part of the split of the previous mega PR. This moves the query plan building and query execution strategy.

**The scope:**
This PR finally
allows the Dataset to define multiple storages and the logic to select the storage to run the query onto. This is the StorageQueryPlanBuilder
splits the query execution into two phases: one before storage selection where we allow QueryProcessors defined by the dataset to run; the second after the storage selection where we allow storage specific query processing
adds a StorageQueryPlan abstraction to allow the dataset to provide directions on how to run the query. This is useful for joins (when the directions define how to join multiple storages) or for split query (where we can let a storage define how to split a query).

**The query execution process**
See snuba/web/query for the details. Now the query execution follows these phases:

The query class is built and parsing is done (no change)
Extensions are applied (no change)
dataset query processors are applied (like before, these are the processors to be appleid on the snuba query)
StorageQueryPlanBuilder (provided by the dataset) prepares the plan by selecting the storage/s and providing a QueryPlanExecutionStrategy to run the query
QueryPlanExecutionStrategy orchestrate the execution of the query being passed a runner function that knows how to query the DB. This is the place where split can be done (in another PR in the stack)

**Limitations**
This does not introduce Entities. It follows the design discussed so Entities can be added on top
Joins are not properly supported. There is a PR on top of this that fixes joins
Splits are moved into storages in a PR provided later in the stack
This does not split the Query object between SnubaQuery and ClickhouseQuery. That needs the AST to work.
This still does not provide a way to univocally identify a storage inside a dataset. This means, given a dataset there is no way to retrieve one specific Storage (they don't have identifiers). This would be useful to make the the consumer not to blindly depend on a dataset name. This would be an easy addition but it would require decisions on how to identify storages.

**On entities**
Entities are not in the scope of this change, my idea on how to fit them in the model is that they will sit in between dataset and storage (this should not be a unexpected).
The model would be dataset --(1:n)--> entity --(1:n)--> storage. So the change to the data structures would be verbose but not very complex.
Entities should be reusable between datasets as we do for Storages.
Query processing would add an additional phase between query parsing and StorageQueryPlanBuilder. This includes resolving the entity (or entities in case of joins) involved in the query, potentially run Entity specific query processors (if we introduce them) and then pass the ball to StorageQueryPlanBuilder.

**Instructions to review**
I feel it is hard to get the context of how the pieces work together if it is split bottom up.
I would look at the PR in these steps:

Look first at _run_query in web/query.py. This puts everything together so it shows how the whole process works
Look at the Storage interface in snuba/datasets/storage.py. This is quite documented [you did this in the PR before]
See how we define storages in a single table dataset (events) and in discover which now should look a lot better.
Look into the Execution plan interface in snuba/datasets/plans/query_plan.py

Depends on PR #828

### CODE REVIEW

File: snuba/datasets/plans/single_table.py [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+
+    def __init__(
+        self, selector: QueryStorageSelector, post_processors: Sequence[QueryProcessor]
+    ) -> None:
+        self.__selector = selector
+        self.__post_processor = post_processors
+
+    def build_plan(self, request: Request) -> StorageQueryPlan:
+        storage = self.__selector.select_storage(request.query, request.settings)
+        return SingleTableQueryPlanBuilder(storage, self.__post_processor,).build_plan(
```
Comment (tkaemming) [inline snuba/datasets/plans/single_table.py]:
```suggestion
        return SingleTableQueryPlanBuilder(storage, self.__post_processor).build_plan(
```

File: snuba/datasets/plans/single_table.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        self.__post_processors = post_processors
+
+    def build_plan(self, request: Request) -> StorageQueryPlan:
+        request.query.set_data_source(
+            self.__storage.get_schemas().get_read_schema().get_data_source()
+        )
+
+        return StorageQueryPlan(
+            query_processors=list(self.__storage.get_query_processors())
+            + list(self.__post_processors),
```
Comment (tkaemming) [inline snuba/datasets/plans/single_table.py]: This is going to presumably create three lists, I'd assume that `[*a, *b]` will only create one and will also format better.

File: snuba/datasets/plans/single_table.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    ) -> None:
+        # The storage the query is based on
+        self.__storage = storage
+        # This is a set of query processors that have to be executed on the
+        # query after the storage selection but that are defined by the dataset.
+        # Query processors defined by a Storage must be executable independently
+        # from the context the Storage is used (whether the storage is used by
+        # itself or whether it is joined with another storage).
+        # In a joined query we would have processors defined by multiple storages.
+        # that would have to be executed only once (liek Prewhere). That is a
```
Comment (tkaemming) [inline snuba/datasets/plans/single_table.py]:
```suggestion
        # that would have to be executed only once (like Prewhere). That is a
```

File: snuba/datasets/plans/query_plan.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+from abc import ABC
+from dataclasses import dataclass
+from typing import Callable, Sequence
+
+from snuba.query import RawQueryResult
+from snuba.query.query_processor import QueryProcessor
+from snuba.request import Request
+
+
+SingleQueryRunner = Callable[[Request], RawQueryResult]
```
Comment (tkaemming) [inline snuba/datasets/plans/query_plan.py]: What does "Single" mean here (or in other words, why isn't this just called "QueryRunner"?)

**Review by tkaemming — CHANGES_REQUESTED** [review]
Some minor things inline, but biggest issue with this that I have is that I pretty strongly disagree with moving the `PrewhereProcessor` to the plan builder for the reasons noted inline. Let me know if there is some reason that has to be there that I am not seeing.

File: snuba/datasets/cdc/groupassignee.py [hunk 5]
```diff
@@ -79,14 +80,14 @@ def __init__(self) -> None:
...
         )
 
-        storage_selector = SingleStorageSelector(storage=storage)
-
         super().__init__(
             storages=[storage],
-            storage_selector=storage_selector,
+            query_plan_builder=SingleTableQueryPlanBuilder(
+                storage=storage, post_processors=[PrewhereProcessor()],
+            ),
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/datasets/cdc/groupassignee.py]: The issue we have with adding the prewhere processing in the individual storage is that it may not work for joins.
Let's say we have a join between two storages (since one storage is one table), each of them has its own prewhere candidates. We could execute the prewhere processor individually, and each one would move some conditions from the where clause to the pre-where.
In a join, we want to apply the pre-where conditions only to the most external table (I suspect it would not even work otherwise) and in general we want to cap the number of conditions we add to the prewhere clause.
This means the prewhere clause cannot be fully processed at storage level in isolation. 
What we could do is to apply that processing on the individual storages, and then reprocessing the clause in the planner to adjust it or remove individual conditions.

Right now it would work either way since the PrewhereProcessor takes the list of candidates from the query `data source` which is populated after the storage is selected, so it would work on SelectedTableQueryPlanBuilder as well (https://github.com/getsentry/snuba/blob/master/snuba/query/processors/prewhere.py#L26).

I am open to move this back to the storage and add a Prewhere-re-processor in the planner to adjust what the storages have decided (which essentially would do nothing in all cases but joins). What do you think.

File: snuba/datasets/groups.py [hunk 6]
```diff
@@ -132,7 +122,9 @@ def __init__(self) -> None:
         storage = JoinedStorage(join_structure)
         super().__init__(
             storages=[storage],
-            storage_selector=GroupsQueryStorageSelector(storage),
+            query_plan_builder=SingleTableQueryPlanBuilder(
+                storage=storage, post_processors=[PrewhereProcessor()],
```
Comment (lynnagara) [inline snuba/datasets/groups.py]: Why do we pass a joined storage into a `SingleTableQueryPlanBuilder` here? The name implies to me that it should only be building queries for single tables and not joins.

Comment (fpacifici) [reply snuba/datasets/groups.py]: The only reason is that the joined storage disappears in PR #830, then there will be a JoinedPlanner that takes multiple storages to join and no JoinedStorage anymore. 
This intermediate step is needed to split that into a separate PR, I can rename it to SingleStorageQueryPlanBuilder like I did for the previous PR (SingleStorageSelector).

File: snuba/query/__init__.py [hunk 7]
```diff
@@ -0,0 +1,8 @@
+from typing import Any, NamedTuple
+
+from snuba.reader import Result
+
+
+class RawQueryResult(NamedTuple):
+    result: Result
+    extra: Any
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/query/__init__.py]: Not sure I get the point.
Do you mean that RawQueryResult (which is the return type of the lowest level function we have that runs the clickhouse query) should be coupled with the web package?
Or that the return type of the function that performs the clickhouse query should NOT be coupled with the web package but it is now ?
If it is the first, what's the reasoning ? I would say the code that runs the clickhouse query does not need to be strictly coupled with the web module otherwise we could execute queries only through the web module.

File: snuba/web/query.py [hunk 8]
```diff
@@ -285,44 +280,14 @@ def parse_and_run_query(
     return result
 
 
-@split_query
-def _run_query(
+def _run_clickhouse_query(
```
Comment (tkaemming) [inline snuba/web/query.py]: Most of this function is not ClickHouse specific… we earlier were trying to make as much of this storage agnostic as possible but this implies we're going the other way?

File: snuba/query/__init__.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/query/__init__.py]: > Do you mean that RawQueryResult (which is the return type of the lowest level function we have that runs the clickhouse query) should be coupled with the web package?

This isn't the lowest level function. The lowest level function we have that runs the ClickHouse query is `Reader.execute`, which _is_ intentionally decoupled from the web package.

> Or that the return type of the function that performs the clickhouse query should NOT be coupled with the web package but it is now ?

It shouldn't be coupled with the web package — and to jump back into beating my eternally dead horse — in my opinion this type _shouldn't actually exist at all_. The `extra` data provides dubiously little value beyond our own test suite which depends on it for frustrating reasons and implied API contracts with users of the HTTP API. It is also returns unreliable (wrong) results when splitting: https://github.com/getsentry/snuba/blob/ab6b98ca7599c6b223e48c917e9ec6d4df473058/snuba/web/split.py#L91-L93

It's not even useless, it's actively harmful in my opinion since the data is _wrong_ in many common cases.

The introduction of the `Result` type from the reader was supposed to move towards fixing this, but I was never able to make progress on this because of the sheer amount of complexity and coordination involved with removing it.

>  I would say the code that runs the clickhouse query does not need to be strictly coupled with the web module otherwise we could execute queries only through the web module.

That's my point. This is literally true today.

File: snuba/datasets/plans/query_plan.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    to the query after the the storage has been selected.
+    It also provides a plan execution strategy, in case the query is not
+    one individual query statement (like for split queries).
+    """
+
+    # TODO: When we will have a separate Query class for Snuba Query and
+    # Storage QUery, this plan will also provide the Storage Query. Right
+    # now the storage query is the same mutable object referenced by Request
+    # so no need to add an additional reference here (it would make the query
+    # execution code more confusing).
```
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/datasets/plans/query_plan.py]: Oops, forgot to actually come back to this (and the ones below) before submitting. I will follow up on this if there's something worth talking about. Watch this space (or not.)

File: snuba/query/__init__.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/query/__init__.py]: I'm not sure as I'd agree it's fully orthogonal, but I see your point and understand at least the literal reasoning for moving this to avoid the circular import (assuming it exists.)

My issue is that moving it here implies that all of the extra data — data that solely exists to facilitate an existing implicit contract made by the HTTP response body structure — is now defined outside of the web module. To anyone who reads this code, this implies that the data has a use outside of the web context. It doesn't.[1] This is not an orthogonal change in my perspective. Non-functional, sure.

Like in another thread in this review, we can't just continue to move code around for the sake of expediency and assume that some reader at a later date will assume the intentions of the author were to not make that a permanent  and intentional change. I'd like to think that we already learned this through the confusion around the events, transactions, and discover dataset subscriptions design and all of the questions and discussions that came out of that. I don't know of any actual plans to fix this, so this is just going to sit there as-is until it the next person who comes along and reads it assumes it was put there for a reason and not just to resolve a circular import problem somewhere else.

I guess I see several options here:

1. leave this as-is and assume anyone reading this later was either part of this conversation, or was warned by someone who was about this before they draw their own conclusions about the intention of the data structure
2. stop working on this, remove the stats from the response, fix the scope of the tests that assert on raw SQL or stats values in the response body, obviate the need this data structure entirely
3. add a comment explaining that this data structure only exists to allow compatibility with the existing web response format even though it doesn't live in the web package and that it should be removed in the future[2]

1: I'm assuming we agree/agreed about this. Maybe we didn't.
2: As an example, my assumption was that subscriptions importing from snuba.web made this obviously clear to anyone familiar with the execution model of subscriptions without needing to specifically call it out. I'd assume the same of query plans. Maybe I'm being too clever.

[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/query/__init__.py]: I think I disagree with several parts of the comment, though I believe the culprit of the issue (which covers most of them) is that we have a different understanding on the role of the `extra` field. 
I personally find it useful, I think sending debugging information back to the user when developing is important. I use debug mode a lot and actually I would have added more to that as the query processing becomes more complex with query planners and multiple query translation steps and query processors. I don't think it is just a short cut for our bad approach at testing (that should be fixed but it does not remove the need for returning debug information). 

As a consequence I did not see moving RawQueryResult down as an expedient (otherwise I would have just broken it off from web/query and put it in web/types or something like that to quickly solve the circular dependency).
When I saw it in the web module, it was a no brainer for me to think it was not supposed not it needed to be there. Not that it was inherently tied to the web interface. I did see subscriptions depending on web as a short cut to make those PRs smaller (I reviewed it as well), but I did not see a reason why that should apply to the RawQueryResult object.
Regarding the wrong results for split queries. It is true, but I prefer fixing it over dropping the feature altogether, in fact this stack allows us to fix it since it pushes the split down to the storage layer (in the last pr) which means we can process the results properly.

Now, about the implicit coupling with the web interface. I see it in this data structure, but I think that's a byproduct of a data structure that is not properly typed and under specified, not because of its existence.
Let's imagine this scenario:
let's pretend that, instead of an untyped `extra` dictionary, it was a dataclass `DebugInfo` with each specific field properly typed and only filled in in debug mode. At the end of the query, part of this content would be written on the querylog kafka topic. Part would be just logged (if snuba was in debug mode) and part was returned to the client if the client was in debug mode. This does not seem that different than how many DB engine work. Would you still say that the layer that is producing the stats should not be allowed to return them back to the client? If yes, would you have an alternative on how to provide debug information back?

[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/query/__init__.py]: **Specifics**

> I personally find it useful

To be clear, I don't disagree that having better visibility into what is happening during query processing is useful. I think the query log is a very good thing for similar reasons.

I personally don't find the current implementation (as reflected in this return type and the web response) very useful though — at least in part since when this kind of data ceases to reflect reality, I stop paying attention to it since I no longer trust it.

I also keep merging "stats" and "timings" together here as well, which doesn't necessarily apply in this specific scenario but is probably relevant when talking about the web response format.

> I think sending debugging information back to the user when developing is important. 

I'm assuming "the user" is really the developer who is _working_ on the "database" software itself (locally or not) — and not the user who is _using_ the database (as a black box), right? Just trying to me extremely clear about who we're talking about as the audience for this data: in other words — it's just us, right?

> our bad approach at testing (that should be fixed)

At least we agree on that part. 🙂

---

**Generalities**

I don't really think this is useful as part of the web response body for a few reasons, roughly ranked in order of importance (in my opinion):

1. It's wrong (or at least not guaranteed to be right in all cases.)
2. This data already should exist in the query log (or additionally in Sentry APM/tracing itself, ignoring the whole ouroboros thing.)
    1. My feeling is that the response structure between the web and subscriptions should be the same (it is not today, intentionally, per #792.) Assuming the results are structured and encoded the same, if debug is enabled on this means we're writing that data to the result stream twice. I think debug is _also_ useful (in theory) for those queries, and eventually we're going to need to be able to provide something along the lines of random access reads for debug data by query ID anyway to get debug information for _those_ specific needles in the haystack.
3. It is unclear to me how this would be properly merged during a split query. (I don't know if I should be interpreting the comment above as "this _is_ fixed in a later change" or "this _could be_ fixed in a later change", so if there is a specific direction already moving here, please let me know.)
4. It prevents (or at least complicates) streaming responses (for things like CSV export or progressively returning rows during a time split.) In theory, we could "easily" support streaming responses by changing the `Result` type to be include an `Iterator[Row]` rather than `Sequence[Row]` and making some "relatively minor" changes to the output codec used for web requests so that each row is emitted as a single line using chunked transfer encoding. I do not see where debug information fits there cleanly.
5. It embeds a cross-cutting concern in the response data structure.

---

**Constructive Feedback**

>  If yes, would you have an alternative on how to provide debug information back?

How I have done this in the past (we instrumented a similar system at Disqus, and this is somewhat similar to the trace data logged by service delegator during the Snuba development and cutover):

Query Path:

1. Assign a unique request ID to each Snuba query. This already happens: https://github.com/getsentry/snuba/blob/master/snuba/request/__init__.py#L14
2. Collect debug information during all stages of request processing in a structure like you mentioned. The structure itself should be reflective of the ordering and nesting processing steps like the output of `EXPLAIN ANALYZE`. (There's probably some opportunity for consistency here with the query plan structure itself but I haven't thought that far ahead yet.) This probably contains a lot more information than exists today. This probably necessitates some processor changes.
3. Include the request ID with the response. (HTTP headers for web, probably Kafka headers for subscriptions.)
4. When the request is "finalized" (HTTP request end for web, Kafka publish for subscriptions), write the debug information to the query log topic.

Debug Information Write Path:

Basically the same as designed today in #783 (but it will need the Kafka partition and offset.)

Read (Debug) Path:

1. Provide an endpoint that accepts a request ID.
2. Look up that request ID in ClickHouse to get the Kafka partition and offset of the raw debug dump.
3. Get the debug data by fetching the message at that partition and offset and return it. This also takes care of automatically TTLing debug data.

[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/query/__init__.py]: I think we agree at least on the fact that building debug information along the query execution and having that data structure available when the query is finalized, right? There is a lot of room for discussion on how that would be implemented but that seems a totally independent problem (I would advise you to sync with Lyn, because what she is building is not 100% in line with this vision).

Today the place where the query is finalized is `parse_and_run_query`, which is where we record the query stats in the Kafka topic.
I guess we could push that down to any place after the components of the split query have been reassembled. Which, would be performed by QueryPlanExecutionStrategy at the end of this stack. See https://github.com/getsentry/snuba/pull/831/files#diff-d733b517dee05cd4f92c17a3f40d77e0R38
This, to me, means only two alternatives:
1) if the debug data structure is passed back and forth during the query execution process (from a phase to the next), then the `QueryRunner` function, as well as any other steps in the query, needs to know about that type, which is the only reason I am moving RawQueryResult in this PR. This need will still be there.
2) the debug information is available in some sort of request context everybody has access to. Maybe something that moves along with the request (not the request itself).

If we are aligned on this, it still seems to me that the issue is something I mentioned above:  `_run_query` should not be in the web module. It should have access to the `Result` object and to this debug data structure (whether or not result and debug are wrapped in the same class) and this Debug data structure should not be in the web module. 
Does this make sense ? If yes, it still seems to me the issue is not in moving the RawResult data structure, but in the fact the web module has access to it (specifically extra data).

About your options to move ahead:
option 1) Let's agree to disagree and drop it since there are better options ?
option 2) Considering the amount of open problems that this thread uncovered I would expect that to take months. Considering there is an engineer in the team who is literally working on observability, I'd rather not block this on that work.
option 3) I think I can add that comment but, better, I think we can also easily go around the circular dependency by breaking web/query.py leaving the Response object in the web module so the issue remains visible at least in the same way as it is for subscriptions. What do you think?

[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/query/__init__.py]: moved to snuba/web/__init__ and added the comment. Won't mark as solved because the debate is still relevant.

File: snuba/web/query.py (hunk 8 above)
Comment (fpacifici) [reply snuba/web/query.py]: I tried splitting this file in different ways and in the end  I am not sure I agree with the point that the responsibilities are not well defined at least for most methods. We have:
- run the query with caching deduping and rate limiting
- run the query with stats recording
- run the query with APM instrumentaion
- process query and run. 

Apart from splitting caching, deduping, and rate limiting from actually running the query, I  would not make major changes to the split of responsibilities. I really think we have a vocabulary problem there.

This does not mean there is no opportunity to improve this class, but I don't think the main issue is the split of responsibilities.

Comment (fpacifici) [reply snuba/web/query.py]: Let me try one last change, hold on on this comment.

File: snuba/query/__init__.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/query/__init__.py]: > I think we agree at least on the fact that building debug information along the query execution and having that data structure available when the query is finalized, right?

Yep, on the same page here.

> There is a lot of room for discussion on how that would be implemented but that seems a totally independent problem (I would advise you to sync with Lyn, because what she is building is not 100% in line with this vision).

Makes sense. I would like us to be more proactively communicative about these things. The extent that I know of what is going on is what is in #739 and #783. It's hard to have any constructive input after the work has already been done. (I don't think this is an isolated incident, either.)

> Today the place where the query is finalized is `parse_and_run_query`, which is where we record the query stats in the Kafka topic. I guess we could push that down to any place after the components of the split query have been reassembled. 

I think that `parse_and_run_query` is the wrong place for this for reasons I've articulated in the past and won't go into again here (unless you want/need me to.) I think it must happen after reassembling the split results (and probably after serialization for response, even) to be useful.

Regarding the two alternatives, I think the debug structure needs to be initialized at the top of the stack so that we can start recording information as soon as possible from things like query processors, rather than accumulating data from the bottom up into the result as the stack unwinds.

> If we are aligned on this, it still seems to me that the issue is something I mentioned above: `_run_query` should not be in the web module.

I don't disagree with this in general. I'm not sure whether or not I agree with it based on the current implementation (basically for the same reasons as `RawQueryResult`.) If the implementation changes so that there's less web coupling here (contentious point whether or not this is true, I know) and this function is just used for generic query execution, I absolutely agree it belongs somewhere else.

> It should have access to the `Result` object and to this debug data structure (whether or not result and debug are wrapped in the same class) and this Debug data structure should not be in the web module. Does this make sense ? 

Yes, 100%.

> If yes, it still seems to me the issue is not in moving the RawResult data structure, but in the fact the web module has access to it (specifically extra data).

Not sure I understand or agree with this point (again, as tied to the current implementation — I think changing this implementation makes this a non-issue by likely removing this data structure.)

> option 1) Let's agree to disagree and drop it since there are better options ?

👍 

> option 2) Considering the amount of open problems that this thread uncovered I would expect that to take months. Considering there is an engineer in the team who is literally working on observability, I'd rather not block this on that work.

I agree that it doesn't really make sense to block this work on that work. (I knowingly wrote both of these two options understanding that they are unrealistic solutions right now, if that was not clear.)

I just want to see us making progress on solving these problems (even acknowledging these are problems and having a conversation like this is progress in my opinion) rather than pushing code around to avoid them, especially as something as important as this part of query execution. If we're interacting with it, we don't have to necessarily fix it, but I do think we need to make marginal improvements where we can (even if those improvements are just deciding and recording some agreement what this data structure is actually supposed to represent.)

> option 3) I think I can add that comment but, better, I think we can also easily go around the circular dependency by breaking web/query.py leaving the Response object in the web module so the issue remains visible at least in the same way as it is for subscriptions. What do you think?

I think this is fine for now, thank you. I think that especially as/if we grow the team (or if we end up looking back at this months later), explicitly calling this out as "something that we know isn't quite right but we haven't figured out a plan to address it" will avoid more misunderstandings down the line (or at least put a minefield warning where the differences of opinion do already exist.)

File: snuba/datasets/cdc/groupassignee.py [hunk 10]
```diff
@@ -82,20 +83,17 @@ def __init__(self) -> None:
...
         )
 
-        storage_selector = SingleStorageSelector(storage=storage)
-
         super().__init__(
             storages=[storage],
-            storage_selector=storage_selector,
+            query_plan_builder=SingleTableQueryPlanBuilder(
+                storage=storage, post_processors=[],
+            ),
```
Comment (tkaemming) [inline snuba/datasets/cdc/groupassignee.py]: Should `post_processors` have a default value if it commonly will be an empty list? Or is there value in explicitly calling out the lack of post-processors?

File: snuba/datasets/dataset.py [hunk 11]
```diff
@@ -82,7 +85,7 @@ def get_extensions(self) -> Mapping[str, QueryExtension]:
     def get_query_processors(self) -> Sequence[QueryProcessor]:
         """
         Returns a series of transformation functions (in the form of QueryProcessor objects)
-        that are applied to queries after parsing and before running them on Clickhouse.
```
Comment (tkaemming) [inline snuba/datasets/dataset.py]: This is a good clarification, thanks

File: snuba/datasets/plans/query_plan.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+    against the DB.
+    Potentially this could be agnostic to the DB.
+    """
+
+    @abstractmethod
+    def execute(self, request: Request, runner: QueryRunner) -> RawQueryResult:
+        """
+        Executes the query plan. The request parameter provides query and query settings.
+        The runner parameters is a function to actually run one individual query on the
+        database.
```
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/datasets/plans/query_plan.py]: (This is probably something that will make more sense to me in context, marking this as resolved for now.)

**Review by tkaemming — APPROVED** [review]
This looks good, whether or not you want to tackle https://github.com/getsentry/snuba/pull/840#discussion_r395700174 as part of this or separately is up to you, I think this is an improvement regardless.

File: snuba/web/query.py (hunk 8 above)
Comment (fpacifici) [reply snuba/web/query.py]: I am about to send a separate PR based on this. I would not be comfortable in deploying this change and the one I am making together, too risky.

File: snuba/datasets/plans/single_storage.py [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+# depend on Result + some debug data structure instead. Also It requires removing
+# extra data from the result of the query.
+from snuba.web import RawQueryResult
+from snuba.query.query_processor import QueryProcessor
+from snuba.request import Request
+
+
+class SimpleQueryPlanExecutionStrategy(QueryPlanExecutionStrategy):
+    def execute(self, request: Request, runner: QueryRunner) -> RawQueryResult:
+        return runner(request)
```
Comment (lynnagara) [inline snuba/datasets/plans/single_storage.py]: This doesn't do much yet. What are the planned future responsibilities of QueryPlanExecutionStrategy? Where will splitting be?

Comment (fpacifici) [reply snuba/datasets/plans/single_storage.py]: splitting will be in another class that extends QUeryPlanExecutionStrategy.
see #831
